### PR TITLE
[WIP] add import/export of products

### PR DESCRIPTION
**Pull request by oallain**
Issue #4 
Import and Export Products datas:

- 'Code'
- 'Name'
- 'Description'
- 'Short_description'
- 'Meta_description'
- 'Meta_keywords'
- 'Main_taxon'
- Attributes (one column by attribute)

TO DO:

- [x] Import - Update product details
- [x] Import - Update product mainTaxon
- [x] Import - Update product attributes
- [x] Import - Create product
- [x] Export - product details
- [x] Export - product mainTaxon
- [x] Export - product attributes (1 column by attribute)
- [x] Behat - test export UI
- [x] Behat - test export CLI
- [x] Behat - test import UI
- [x] Behat - test import CLI
- [x] README - add doc

### CODE REVIEW

File: src/Resources/config/services_export_csv.yml [hunk 1]
```diff
@@ -1,6 +1,8 @@
 services:
     sylius.csv_writer:
         class: Port\Csv\CsvWriter
+        arguments:
+            - ","
```
Comment (oallain) [inline src/Resources/config/services_export_csv.yml]: I changed this value to be able to export and re-import the same file.

Comment (oallain) [reply src/Resources/config/services_export_csv.yml]: I remove it and make an other PR ;)

File: src/Exporter/Plugin/ProductPluginPool.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+declare(strict_types=1);
+
+namespace FriendsOfSylius\SyliusImportExportPlugin\Exporter\Plugin;
+
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+
+class ProductPluginPool extends PluginPool
```
Comment (stefandoorn) [inline src/Exporter/Plugin/ProductPluginPool.php]: Can be marked final

File: src/Exporter/Plugin/ProductPluginPool.php [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+{
+    public function __construct(array $plugins, array $exportKeys, RepositoryInterface $productAttributeRepository)
+    {
+        $attrSlug = [];
+        $productAttr = $productAttributeRepository->findBy([], ['id' => 'ASC']);
```
Comment (stefandoorn) [inline src/Exporter/Plugin/ProductPluginPool.php]: I don't think we should be performing queries in constructors, isn't performant and can give issues with cache etc. I think this should be done in one of the public methods already in the extended class (`initPlugins` I think is best). Then you do this & call the parent `initPlugins` method.

File: src/Exporter/Plugin/ProductResourcePlugin.php [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+namespace FriendsOfSylius\SyliusImportExportPlugin\Exporter\Plugin;
+
+use Doctrine\ORM\EntityManagerInterface;
+use Sylius\Component\Attribute\Model\AttributeValueInterface;
+use Sylius\Component\Core\Model\ProductInterface;
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+use Symfony\Component\PropertyAccess\PropertyAccessorInterface;
+
+class ProductResourcePlugin extends ResourcePlugin
```
Comment (stefandoorn) [inline src/Exporter/Plugin/ProductResourcePlugin.php]: Can be marked final

File: src/Exporter/Plugin/ProductResourcePlugin.php [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+{
+    public function __construct(
```
Comment (stefandoorn) [inline src/Exporter/Plugin/ProductResourcePlugin.php]: Don't think we need this construct, it looks exactly like the extended constructor, isn't it?

File: src/Exporter/Plugin/ProductResourcePlugin.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function init(array $idsToExport): void
+    {
+        parent::init($idsToExport);
+
+        /** @var ProductInterface $resource */
+        foreach ($this->resources as $resource) {
+            // insert Translation information to specific fields
```
Comment (stefandoorn) [inline src/Exporter/Plugin/ProductResourcePlugin.php]: Comments here can be left out, the method names are descriptive

File: src/Exporter/ProductResourceExporter.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+{
+    public function __construct(
+        WriterInterface $writer,
+        PluginPoolInterface $pluginPool,
+        array $resourceKeys,
+        RepositoryInterface $productAttributeRepository,
+        ?TransformerPoolInterface $transformerPool
+    ) {
+        $attrSlug = [];
+        $productAttr = $productAttributeRepository->findBy([], ['id' => 'ASC']);
```
Comment (stefandoorn) [inline src/Exporter/ProductResourceExporter.php]: Same here, we shouldn't be performing queries in constructors

File: src/Exporter/ProductResourceExporter.php [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+declare(strict_types=1);
+
+namespace FriendsOfSylius\SyliusImportExportPlugin\Exporter;
+
+use FriendsOfSylius\SyliusImportExportPlugin\Exporter\Plugin\PluginPoolInterface;
+use FriendsOfSylius\SyliusImportExportPlugin\Exporter\Transformer\TransformerPoolInterface;
+use FriendsOfSylius\SyliusImportExportPlugin\Writer\WriterInterface;
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+
+class ProductResourceExporter extends ResourceExporter
```
Comment (stefandoorn) [inline src/Exporter/ProductResourceExporter.php]: Can be marked final

File: src/Processor/ProductProcessor.php [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        $this->resourceProductFactory = $productFactory;
+        $this->resourceTaxonFactory = $taxonFactory;
+        $this->productRepository = $productRepository;
+        $this->taxonRepository = $taxonRepository;
+        $this->metadataValidator = $metadataValidator;
+        $this->propertyAccessor = $propertyAccessor;
+        $this->productAttributeRepository = $productAttributeRepository;
+
+        $this->attrCode = [];
+        $productAttr = $productAttributeRepository->findBy([], ['id' => 'ASC']);
```
Comment (stefandoorn) [inline src/Processor/ProductProcessor.php]: Same here, no queries in constructors

File: src/Processor/ProductProcessor.php [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    private function setAttributesData(ProductInterface $product, array $data): void
+    {
+        foreach ($this->attrCode as $attrCode) {
+            if ($product->getAttributeByCodeAndLocale($attrCode) !== null) {
+                $product->getAttributeByCodeAndLocale($attrCode)->setValue($data[$attrCode]);
+
+                continue;
+            }
+
+            if (!empty($data[$attrCode])) {
```
Comment (stefandoorn) [inline src/Processor/ProductProcessor.php]: I would turn this around:

```
if (empty($data[$attrCode])) {
    continue;
}
```

And then the rest. It gives clear intends on when to stop and when to continue without a lot of indentation needed.

File: src/Resources/config/services_import_csv.yml [hunk 11]
```diff
@@ -56,3 +56,17 @@ services:
             - "%sylius.importer.stop_on_failure%"
         tags:
             - { name: sylius.importer, type: tax_category, format: csv }
+
```
Comment (stefandoorn) [inline src/Resources/config/services_import_csv.yml]: Double newline

File: src/Resources/translations/messages.en.yml [hunk 12]
```diff
@@ -8,4 +8,3 @@ fos:
                 csv: 'CSV'
                 xlsx: 'Excel'
                 json: 'JSON'
-
```
Comment (stefandoorn) [inline src/Resources/translations/messages.en.yml]: Unrelated change :-)

**Comment by oallain** [comment]
@stefandoorn Thank you for reviewing.
I will do my best to correct all that and do tests :slightly_smiling_face:

File: src/Exporter/Plugin/ProductPluginPool.php (hunk 3 above)
Comment (oallain) [reply src/Exporter/Plugin/ProductPluginPool.php]: I do that, but i need to switch attribute `exportKeys` from private to protected in PluginPool class ;)

File: src/Processor/ProductProcessor.php [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+        $this->headerKeys = $headerKeys;
+        $this->slugGenerator = $slugGenerator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function process(array $data): void
+    {
+        $this->attrCode = $this->attributesCode->getAttributeCodesList();
```
Comment (solutionDrive-Alt) [inline src/Processor/ProductProcessor.php]: Is there a reason for storing the result of this in a `property`?

File: src/Service/AttributesCode.php [hunk 14]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+declare(strict_types=1);
+
+namespace FriendsOfSylius\SyliusImportExportPlugin\Service;
+
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+
+final class AttributesCode implements AttributesCodeInterface
```
Comment (solutionDrive-Alt) [inline src/Service/AttributesCode.php]: I'm not totally happy with the name of this `class`. This class `provides` a `list` of  `codes` of all `attributes`. So what do you think about `AttributeCodesProvider` or something like that?
The current name does not tell me what it does.

File: src/Processor/ProductProcessor.php [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+    private function setAttributesData(ProductInterface $product, array $data): void
+    {
+        foreach ($this->attrCode as $attrCode) {
+            if ($product->getAttributeByCodeAndLocale($attrCode) !== null) {
+                $product->getAttributeByCodeAndLocale($attrCode)->setValue($data[$attrCode]);
+
+                continue;
+            }
+
+            if (empty($data[$attrCode])) {
```
Comment (solutionDrive-Alt) [inline src/Processor/ProductProcessor.php]: maybe we should test this first, because if this is empty we could save one call on `getAttributeByCodeAndLocale ` which should be more expensive than the test of `empty`, or is there a reason why it is in this order?

File: src/Processor/ProductProcessor.php [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+                continue;
+            }
+
+            $productAttr = $this->productAttributeRepository->findOneBy(['code' => $attrCode]);
```
Comment (solutionDrive-Alt) [inline src/Processor/ProductProcessor.php]: I think this (up to line 131) can be moved to its own method.

File: src/Processor/ProductProcessor.php [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+    ) {
+        $this->resourceProductFactory = $productFactory;
+        $this->resourceTaxonFactory = $taxonFactory;
+        $this->productRepository = $productRepository;
+        $this->taxonRepository = $taxonRepository;
+        $this->metadataValidator = $metadataValidator;
+        $this->propertyAccessor = $propertyAccessor;
+        $this->productAttributeRepository = $productAttributeRepository;
+        $this->productAttributeValueFactory = $productAttributeValueFactory;
+        $this->attributesCode = $attributesCode;
```
Comment (solutionDrive-Alt) [inline src/Processor/ProductProcessor.php]: I saw this injection now a couple of times, maybe we should just inject the result of `getAttributeCodeList` instead of the whole object all the time. 
https://symfony.com/doc/current/service_container/expression_language.html

File: composer.json [hunk 18]
```diff
@@ -5,7 +5,7 @@
     "license": "MIT",
     "require": {
         "php": "^7.1",
-        "sylius/sylius": "^1.0",
+        "sylius/sylius": "1.2.* <1.2.11",
```
[the remark this answers is not in the document]
Comment (oallain) [reply composer.json]: It was a test ;)

File: src/Service/AttributesCode.php (hunk 14 above)
Comment (oallain) [reply src/Service/AttributesCode.php]: Yes, it's better :)

File: src/Processor/ProductProcessor.php (hunk 15 above)
Comment (oallain) [reply src/Processor/ProductProcessor.php]: Yes :+1:

File: src/Processor/ProductProcessor.php (hunk 13 above)
Comment (oallain) [reply src/Processor/ProductProcessor.php]: I use it in `setAttributesData()` too.

File: src/Processor/ProductProcessor.php (hunk 16 above)
Comment (oallain) [reply src/Processor/ProductProcessor.php]: I havn't understand :(

File: src/Processor/ProductProcessor.php (hunk 17 above)
Comment (oallain) [reply src/Processor/ProductProcessor.php]: Are you sure ?
Because my function contains a query, I don't think it's a good thing.

**Comment by oallain** [comment]
Someone knows how to solve the error generated by Liip? I didn't succeed :(

File: src/Processor/ProductProcessor.php (hunk 17 above)
Comment (solutionDrive-Alt) [reply src/Processor/ProductProcessor.php]: Not entirely, but I'm not sure where the problem would be, because this query is rather static and has no dynamic input or something like that. And the result should not change as far as I can see it.
But this is no showstopper for me, it was mostly a thought when seeing this injection in several places combined with the call of the same method.

File: src/Processor/ProductProcessor.php (hunk 16 above)
Comment (solutionDrive-Alt) [reply src/Processor/ProductProcessor.php]: Also no showstopper for me. What I tried to say is, that this part where you do stuff with the `$productAttr` (set values and so on) could be moved to a privat message, to reduce the size of this method.

**Comment by solutionDrive-Alt** [comment]
This build is kind of weird. Half of the jobs fail with path-issues and the other half errors with 
```
Error: Problem finding locally-installed `gulp` for gulpfile undefined. (Try running `npm install gulp` from /home/travis/build/FriendsOfSylius/SyliusImportExportPlugin/vendor/sylius/sylius/src/Sylius/Bundle/AdminBundle to install a local gulp for said gulpfile.)
```
The master build went quite well. At the moment I have no idea what causes this. 
@lsmith77 @stefandoorn Do you have any ideas on that?

**Comment by stefandoorn** [comment]
Not directly @solutionDrive-Alt ..

File: src/Processor/ProductProcessor.php (hunk 16 above)
Comment (oallain) [reply src/Processor/ProductProcessor.php]: Yes, of course, i refacto that !=

**Comment by oallain** [comment]
:tada: :smile: :tada:

**Comment by oallain** [comment]
Hello @solutionDrive-Alt @stefandoorn,

I did my best to finalize this PR.
If you have any comments I will try to deal with them :)

In the near future, I would surely evolve this base to add more product informations.

**Comment by oallain** [comment]
Rebase to master :)

File: features/export/ui/exporting_products.feature [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+
+  @ui
+  Scenario: Exporting products should export all of them
+    When I open the product admin index page
+    And I import product data from "products.csv" csv file
+    Then I should see a notification that the import was successful
+    And I should see 2 products in the list
+    Then I go to "/admin/export/sylius.product/csv" homepage
+    And response should contain "Code,Name,Description,Short_description,Meta_description,Meta_keywords,Main_taxon"
+    And response should contain '123456,"Product 1","Description 1","Short description 1","Meta description 1","Meta keywords 1",'
```
Comment (stefandoorn) [inline features/export/ui/exporting_products.feature]: Can we also check for the attributes that should be included? That's a nice feature and important not to break.

File: src/Exporter/Plugin/ProductPluginPool.php [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+
+declare(strict_types=1);
+
+namespace FriendsOfSylius\SyliusImportExportPlugin\Exporter\Plugin;
+
+use FriendsOfSylius\SyliusImportExportPlugin\Service\AttributeCodesProviderInterface;
+
+final class ProductPluginPool extends PluginPool
+{
+    /** @var \FriendsOfSylius\SyliusImportExportPlugin\Service\AttributeCodesProviderInterface */
```
Comment (stefandoorn) [inline src/Exporter/Plugin/ProductPluginPool.php]: Here the short version can be used, it's already included with the `use` statement

File: src/Exporter/Plugin/ProductPluginPool.php [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+        array $exportKeys,
+        AttributeCodesProviderInterface $attributeCodesProvider
+    ) {
+        parent::__construct($plugins, $exportKeys);
+        $this->attributeCodesProvider = $attributeCodesProvider;
+    }
+
+    public function initPlugins(array $ids): void
+    {
+        $this->exportKeys = \array_merge($this->exportKeys, $this->attributeCodesProvider->getAttributeCodesList());
```
Comment (stefandoorn) [inline src/Exporter/Plugin/ProductPluginPool.php]: Should we also do something with `exportKeysAvailable`?

File: src/Exporter/Plugin/ProductResourcePlugin.php [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function init(array $idsToExport): void
+    {
+        parent::init($idsToExport);
+
+        /** @var ProductInterface $resource */
+        foreach ($this->resources as $resource) {
+            $this->addTranslationData($resource);
```
Comment (stefandoorn) [inline src/Exporter/Plugin/ProductResourcePlugin.php]: This is going to be a performance issue, but the plugin has multiple of them right now by iterating over all results. Later we should provide all this data at once via a repository method.

File: src/Exporter/Plugin/ProductResourcePlugin.php [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+        $translation = $resource->getTranslation();
+
+        $this->addDataForResource($resource, 'Name', $translation->getName());
+        $this->addDataForResource($resource, 'Description', $translation->getDescription());
+        $this->addDataForResource($resource, 'Short_description', $translation->getShortDescription());
+        $this->addDataForResource($resource, 'Meta_description', $translation->getMetaDescription());
+        $this->addDataForResource($resource, 'Meta_keywords', $translation->getMetaKeywords());
+    }
+
+    private function addTaxonData(ProductInterface $resource)
```
Comment (stefandoorn) [inline src/Exporter/Plugin/ProductResourcePlugin.php]: Missing return type

File: src/Exporter/Plugin/ProductResourcePlugin.php [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+        /** @var \Sylius\Component\Core\Model\TaxonInterface $taxon */
+        $taxon = $resource->getMainTaxon();
+        if ($taxon !== null) {
+            $taxonSlug = $taxon->getSlug();
+        }
+
+        $this->addDataForResource($resource, 'Main_taxon', $taxonSlug);
+    }
+
+    private function addAttributeData(ProductInterface $resource)
```
Comment (stefandoorn) [inline src/Exporter/Plugin/ProductResourcePlugin.php]: Missing return type

File: src/Exporter/Plugin/ProductResourcePlugin.php [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+        $this->addDataForResource($resource, 'Meta_keywords', $translation->getMetaKeywords());
+    }
+
+    private function addTaxonData(ProductInterface $resource)
+    {
+        $taxonSlug = '';
+
+        /** @var \Sylius\Component\Core\Model\TaxonInterface $taxon */
+        $taxon = $resource->getMainTaxon();
+        if ($taxon !== null) {
```
Comment (stefandoorn) [inline src/Exporter/Plugin/ProductResourcePlugin.php]: I prefer `if (null !== $taxon) {` syntax

File: src/Processor/ProductProcessor.php [hunk 26]
```diff
@@ -0,0 +1,151 @@
...
+        if (null === $product) {
+            /** @var ProductInterface $product */
+            $product = $this->resourceProductFactory->createNew();
+            $product->setCode($code);
+        }
+
+        return $product;
+    }
+
+    private function setMainTaxon(ProductInterface $product, array $data)
```
Comment (stefandoorn) [inline src/Processor/ProductProcessor.php]: Missing return type

File: src/Processor/ProductProcessor.php [hunk 27]
```diff
@@ -0,0 +1,151 @@
...
+    {
+        /** @var TaxonInterface|null $mainTaxon */
+        $mainTaxon = $this->taxonRepository->findOneBy(['code' => $data['Main_taxon']]);
```
Comment (stefandoorn) [inline src/Processor/ProductProcessor.php]: Is there a need to store the result in a variable first?

File: src/Processor/ProductProcessor.php [hunk 28]
```diff
@@ -0,0 +1,151 @@
...
+                $product->getAttributeByCodeAndLocale($attrCode)->setValue($data[$attrCode]);
+
+                continue;
+            }
+
+            $this->setAttributeValue($product, $data, $attrCode);
+        }
+    }
+
+    private function setDetails(ProductInterface $product, array $data)
```
Comment (stefandoorn) [inline src/Processor/ProductProcessor.php]: Missing return type

File: src/Processor/ProductProcessor.php [hunk 29]
```diff
@@ -0,0 +1,151 @@
...
+    {
+        $product->setName($data['Name']);
+        $product->setDescription($data['Description']);
+        $product->setShortDescription($data['Short_description']);
+        $product->setMetaDescription($data['Meta_description']);
+        $product->setMetaKeywords($data['Meta_keywords']);
+        $product->setSlug($product->getSlug() ?: $this->slugGenerator->generate($product->getName()));
+    }
+
+    private function setAttributeValue(ProductInterface $product, array $data, $attrCode)
```
Comment (stefandoorn) [inline src/Processor/ProductProcessor.php]: Missing return type
Also what is the type of `$attrCode`?

File: src/Service/AttributeCodesProviderInterface.php [hunk 30]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+declare(strict_types=1);
+
+namespace FriendsOfSylius\SyliusImportExportPlugin\Service;
+
+interface AttributeCodesProviderInterface
+{
+    public function getAttributeCodesList(): array;
```
Comment (stefandoorn) [inline src/Service/AttributeCodesProviderInterface.php]: Maybe have a typehint `@return string[]`. I think codes are always strings.

File: tests/Behat/Context/CliProductsContext.php [hunk 31]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tests\FriendsOfSylius\SyliusImportExportPlugin\Behat\Context;
+
+use Behat\Gherkin\Node\TableNode;
+use PHPUnit\Framework\Assert;
+
+class CliProductsContext extends CliBaseContext
```
Comment (stefandoorn) [inline tests/Behat/Context/CliProductsContext.php]: Can be final

File: tests/Behat/Context/CliProductsContext.php [hunk 32]
```diff
@@ -0,0 +1,22 @@
...
+{
+    /**
+     * @Then I should have at least the following product ids in the database:
+     */
+    public function iShouldHaveAtLeastTheFollowingTaxCategoriesIdsInTheDatabase(TableNode $productIds)
```
Comment (stefandoorn) [inline tests/Behat/Context/CliProductsContext.php]: Missing return type

File: tests/Behat/Context/ProductsContext.php [hunk 33]
```diff
@@ -0,0 +1,159 @@
...
+     */
+    public function getRouteName(): string
+    {
+        return 'sylius_admin_product_index';
+    }
+
+    /**
+     * @When I import product data from :file :format file
+     */
+    public function iImportProductDataFromCsvFile(string $file, string $format)
```
Comment (stefandoorn) [inline tests/Behat/Context/ProductsContext.php]: Missing return type

File: tests/Behat/Context/ProductsContext.php [hunk 34]
```diff
@@ -0,0 +1,159 @@
...
+    {
+        $this->productIndexPage->importData($file, $format);
+    }
+
+    /**
+     * @When I open the product admin index page
+     */
+    public function iOpenTheProductIndexPage()
```
Comment (stefandoorn) [inline tests/Behat/Context/ProductsContext.php]: Missing return type

File: tests/Behat/Context/ProductsContext.php [hunk 35]
```diff
@@ -0,0 +1,159 @@
...
+    {
+        $this->productIndexPage->open();
+    }
+
+    /**
+     * @Then I should see an export button
+     */
+    public function iShouldSeeExportButton()
```
Comment (stefandoorn) [inline tests/Behat/Context/ProductsContext.php]: Missing return type

File: tests/Behat/Context/ProductsContext.php [hunk 36]
```diff
@@ -0,0 +1,159 @@
...
+        Assert::assertEquals(
+            'Export',
+            $this->getElement('export_button_text')->getText()
+        );
+    }
+
+    /**
+     * @Then I click on :element
+     */
+    public function iClickOn($element)
```
Comment (stefandoorn) [inline tests/Behat/Context/ProductsContext.php]: Missing return type

Comment (stefandoorn) [reply tests/Behat/Context/ProductsContext.php]: Also, what is type of `$element`?

File: tests/Behat/Context/ProductsContext.php [hunk 37]
```diff
@@ -0,0 +1,159 @@
...
+        if (!$findName) {
+            throw new ElementNotFoundException($element . ' could not be found');
+        }
+        $findName->click();
+    }
+
+    /**
+     * @Then I should see a link to export products to CSV
+     */
+    public function iShouldSeeExportCSVLink()
```
Comment (stefandoorn) [inline tests/Behat/Context/ProductsContext.php]: Missing return type

File: tests/Behat/Context/ProductsContext.php [hunk 38]
```diff
@@ -0,0 +1,159 @@
...
+        Assert::assertContains(
+            'CSV',
+            $this->getElement('export_links')->find('css', 'a.item')->getText()
+        );
+    }
+
+    /**
+     * @Then the product :product should appear in the registry
+     */
+    public function theProductShouldAppearInTheRegistry($productname)
```
Comment (stefandoorn) [inline tests/Behat/Context/ProductsContext.php]: Missing return type & argument type

File: tests/Behat/Context/ProductsContext.php [hunk 39]
```diff
@@ -0,0 +1,159 @@
...
+        Assert::assertNotNull(
+            $product,
+            sprintf('Product with name "%s" does not exist', $productname)
+        );
+    }
+
+    /**
+     * @Then :amount products should be in the registry
+     */
+    public function amountProductsShouldBeInTheRegistry($amount)
```
Comment (stefandoorn) [inline tests/Behat/Context/ProductsContext.php]: Missing return type & argument type

File: tests/Behat/Context/ProductsContext.php [hunk 40]
```diff
@@ -0,0 +1,159 @@
...
+        return array_merge(parent::getDefinedElements(), [
+            'export_button_text' => '.buttons div.dropdown span.text',
+            'export_links' => '.buttons div.dropdown div.menu',
+        ]);
+    }
+
+    /**
+     * @When I go to :hp homepage
+     */
+    public function goToSpecificHomepage($hp)
```
Comment (stefandoorn) [inline tests/Behat/Context/ProductsContext.php]: Missing return type & argument type

File: tests/Behat/Context/ProductsContext.php [hunk 41]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    /**
+     * Checks that response body contains specific text.
+     *
+     * @param string $text
+     *
+     * @Then response should contain :text
+     */
+    public function theResponseShouldContain($text)
```
Comment (stefandoorn) [inline tests/Behat/Context/ProductsContext.php]: Missing return type & argument type

File: features/export/ui/exporting_products.feature (hunk 19 above)
Comment (oallain) [reply features/export/ui/exporting_products.feature]: For that I need to create attributes. I saw that there was a Behat test to do that `Feature: Adding a new text product attribute`, but I don't know how I can make Behat create one in my test. I don't know enough Behat, can you help me?

File: src/Exporter/Plugin/ProductPluginPool.php (hunk 21 above)
Comment (oallain) [reply src/Exporter/Plugin/ProductPluginPool.php]: I haven't see it, but of course, before `initPlugins()` they must be identical. :+1:

File: src/Exporter/Plugin/ProductResourcePlugin.php (hunk 22 above)
Comment (oallain) [reply src/Exporter/Plugin/ProductResourcePlugin.php]: Yes, if we have a performance issue, we can create custom query and add `select` to prevent lazy loading. If I have this problem, will do it in an upcoming issue ;)

File: src/Processor/ProductProcessor.php (hunk 27 above)
Comment (oallain) [reply src/Processor/ProductProcessor.php]: Not realy, i remove it ;)

**Comment by oallain** [comment]
@stefandoorn Thank you for reviewing.
I have a question for you here: https://github.com/FriendsOfSylius/SyliusImportExportPlugin/pull/170#discussion_r262197593

I think the rest is good :+1:

File: features/export/ui/exporting_products.feature (hunk 19 above)
Comment (stefandoorn) [reply features/export/ui/exporting_products.feature]: One of these lines probably imports some products to the database, but I haven't written them I think. @solutionDrive-Alt do you know where? In that case maybe it can simply be added there.

File: src/Exporter/Plugin/PluginPool.php [hunk 42]
```diff
@@ -7,24 +7,24 @@
...
 
     /**
-     * @var array
+     * @var PluginInterface[]
      */
-    private $exportKeysNotFound;
+    private $plugins;
 
     /**
      * @var array
```
Comment (vasilvestre) [inline src/Exporter/Plugin/PluginPool.php]: Typehint array without ? mean it shouldn't be null at any moment. No ?

File: src/Exporter/Plugin/ProductResourcePlugin.php (hunk 22 above)
Comment (vasilvestre) [reply src/Exporter/Plugin/ProductResourcePlugin.php]: Using yield in your method will help a lot if you _clear from entity manager from repository side then

File: src/Exporter/Plugin/PluginPool.php (hunk 42 above)
Comment (oallain) [reply src/Exporter/Plugin/PluginPool.php]: Hello @vasilvestre,

I didn't change this behavior, I just changed the order of attributes here. You can see the original code here: https://github.com/FriendsOfSylius/SyliusImportExportPlugin/pull/170/files/cbb04182927a83fa1e74e7d5f66e6c9b2efbb408#diff-9111ba278b5ac663dded9ff8fdc97398L22

That being said, the attribute is initialized in the constructor as an array and we can't change it later.
It's just ?

File: features/export/ui/exporting_products.feature (hunk 19 above)
Comment (solutionDrive-Alt) [reply features/export/ui/exporting_products.feature]: Our behat tests are not the best, I want to rework them for some time.
But basically Sylius should provide steps to accomplish that.

First you should find a feature within Sylius which has a step like
`there is a product attribute` or something like this. Often it is also something like `this product has an attribute`. Sadly I didn't find it right away and would also have to dig a bit deeper, what I can do in the next days.
If you should find the right step you can use it in your `feature` if the `Context`-Class which holds the implementation of this step is include in the correct `suite` which are definend within  SyliusImportExportPlugin/tests/Behat/Resources/suites.yml. (you would have to find out the service-id of this Context-Class)

You also can implement such a step relatively easy using the sylius-repository/-factory classes and the `sharedStorage` in our own Context-Classes. 

Behat is not trivial and the Behat-Suite of Sylius is rather sophisticated. I plan on refactoring them in this repository in the near future.

Comment (solutionDrive-Alt) [reply features/export/ui/exporting_products.feature]: @stefandoorn @lsmith77  I would suggest, that we define requirements for a behat feature that will ensure the expected behaviour and we implement this feature asap before we merge this to master. Maybe we can create some branch like `feature/product-import-export` which serves as base branch, so we can merge this PR without this feature. WDYT?

The reason why I think we shouldn't do this in this PR is, that we have no exact guideline on how Behat-Tests can be written (this is something I want to work on right now) within our Plugin.
Therefor it is very hard for somebody with little Behat-experience to take the correct steps to get this done. And as I understand it @oallain struggles with behat at the moment.

File: src/Processor/ProductProcessor.php [hunk 43]
```diff
@@ -0,0 +1,148 @@
...
+    {
+        $productAttr = $this->productAttributeRepository->findOneBy(['code' => $attrCode]);
+        /** @var ProductAttributeValueInterface $attr */
+        $attr = $this->productAttributeValueFactory->createNew();
+        $attr->setAttribute($productAttr);
+        $attr->setProduct($product);
+        $attr->setLocaleCode($product->getTranslation()->getLocale());
+        $attr->setValue($data[$attrCode]);
+        $product->addAttribute($attr);
+        $this->productRepository->add($attr);
```
Comment (solutionDrive-Alt) [inline src/Processor/ProductProcessor.php]: shouldn't this be 
```
$this->productAttributeRepository->add($attr);
```
?

File: features/export/ui/exporting_products.feature (hunk 19 above)
Comment (oallain) [reply features/export/ui/exporting_products.feature]: Yeah sorry, i'm a Behat noob :smiley:

File: src/Processor/ProductProcessor.php (hunk 43 above)
Comment (oallain) [reply src/Processor/ProductProcessor.php]: :+1:  Well seen ! Thks

File: src/Exporter/Plugin/ProductResourcePlugin.php (hunk 22 above)
Comment (oallain) [reply src/Exporter/Plugin/ProductResourcePlugin.php]: Yes @vasilvestre , i think too.
We can make it in :

- \FriendsOfSylius\SyliusImportExportPlugin\Exporter\Plugin\ResourcePlugin::init
- \FriendsOfSylius\SyliusImportExportPlugin\Exporter\Plugin\OrderResourcePlugin::init
- \FriendsOfSylius\SyliusImportExportPlugin\Exporter\Plugin\ProductResourcePlugin::init

We make it in an other PR ?

**Comment by oallain** [comment]
> Minor feedback about the code. Except for the behat tests, it looks good to me. For behat we will need some better guidline so everyone working on this plugin can get an idea of how certain things can be achieved.

@solutionDrive-Alt  You will create issue for that ?

File: features/export/ui/exporting_products.feature (hunk 19 above)
Comment (lsmith77) [reply features/export/ui/exporting_products.feature]: @pamil can you maybe help @oallain wrap this up? I assume it would be useful for your current project as well?

File: features/export/ui/exporting_products_with_attributs.feature [hunk 44]
```diff
@@ -0,0 +1,20 @@
...
+  In order to have my products exported to an external target
+  As an Administrator
+  I want to be able to export product data to csv file from backOffice
+
+  Background:
+    Given I am logged in as an administrator
+
+  @ui
+  Scenario: Exporting products should export all of them
+#    Given the store has a text product attribute "T-shirt cotton brand"
```
Comment (oallain) [inline features/export/ui/exporting_products_with_attributs.feature]:
```suggestion

```

File: features/export/ui/exporting_products.feature (hunk 19 above)
Comment (oallain) [reply features/export/ui/exporting_products.feature]: @solutionDrive-Alt, @lsmith77,

I have add another test for product with attributes here : https://github.com/FriendsOfSylius/SyliusImportExportPlugin/pull/170/files#diff-a49fa971491441d0fd6dbc8557aa93a1R2
(After add products/attributes with Behat :muscle: )

File: src/Exporter/Plugin/ProductResourcePlugin.php (hunk 22 above)
Comment (oallain) [reply src/Exporter/Plugin/ProductResourcePlugin.php]: I had create issue https://github.com/FriendsOfSylius/SyliusImportExportPlugin/issues/191 for that

**Comment by oallain** [comment]
This feature need more work ;)
